images/base-devel/debian-pkgs.sh

```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+
+PKGS="binutils binutils-common binutils-x86-64-linux-gnu clang clang-11 cmake cpp cpp-10 curl gcc gcc-10 git lib32gcc-s1 lib32stdc++6 libacl1 libarchive13 libasan6 libatomic1 libbinutils libbrotli1 libbsd0 libbz2-1.0 libc-dev-bin libc6-dev libc6-i386 libcc1-0 libclang-common-11-dev libclang1-11 libclang-cpp11 libcom-err2 libcrypt-dev libctf-nobfd0 libctf0 libcurl4 libedit2 libffi7 libgc1 libgcc-10-dev libgcc-s1 libgcrypt20 libgmp10 libgnutls30 libgomp1 libgpg-error0 libgssapi-krb5-2 libhogweed6 libicu67 libidn2-0 libisl23 libitm1 libjsoncpp24 libk5crypto3 libkeyutils1 libkrb5-3 libkrb5support0 libldap-2.4-2 libllvm11 liblsan0 liblz4-1 liblzma5 libmd0 libmpc3 libmpfr6 libnettle8 libnghttp2-14 libnsl-dev libobjc-10-dev libobjc4 libp11-kit0 libpsl5 libquadmath0 librhash0 librtmp1 libsasl2-2 libssh-4 libssh2-1 libstdc++-10-dev libstdc++6 libtasn1-6 libtinfo6 libtirpc-dev libtsan0 libubsan1 libunistring2 libuv1 libxml2 libz3-4 libzstd1 linux-libc-dev make wget zlib1g"
```

Comment:
The 'tldr;' here is:  let the distro do its job.

I don't think it makes sense to list the vast majority of these packages.

For example, you list "gcc" and "gcc-10".  I would guess the intent here is to either:
a.) have "gcc" (the default version of gcc)
b.) specifically have gcc-10

I don't think you intentionally want to have both. So you should either *only* list 'gcc' (if you want 'b',  *only* list 'gcc-10').  Let the package system do the right thing.

Second, you have listed things like libnghttp2-14 , libtasn1-6, libtinfo6 ... I really doubt you care about the fast majority of lib* packages here. I suspect you wanted 'libcurl4'.  If you "render" the full list here, then there is no way of knowing if you actually wanted to provide libgssapi-krb5-2 or if it just came as a side effect of wanting to provide libcurl.

I suspect your list could boil down to something like:

    binutils clang gcc git libcurl4 libxml2 libzstd1make wget

Reply:
I agree with you @smoser but the installation scripts per design only unupack and copy the content of the packages specified into the image rfs (and results from post install scripts seems to be missing too from the final rfs).

If I understood correctly from @rchincha this was done in order to manually prune everything which wasn't strictly needed for the purpose the package was installed for in the specific context of the image.
The issue I have is I have no way to know what gcc would be used for by developer in order to do that effectively. In case of the rust image, I had figured it out (the current implementation on main branch, not the one in this PR) by compiling a simple rust source, but that is more or less a guess when put in practice in a real world scenario.

Unless we want to update the scripts to also install dependencies of those packages we're stuck listing everything.

FWIW I don't like this system we are using either. Especially since looking at what dockerhub is doing right now in listing installed packages inside the container images (not sure they can do that with distroless though).

Reply:
I think that ignoring package dependencies is not the right thing to be doing.

that said, if i'm reading this right, it looks to me like my suggestion above would both work and probably be significantly faster.

in [defs.sh](https://github.com/project-stacker/c3/blob/ab7da224c682fcd6b38312cf59e16e69bde7a2bb/pkgmgr/defs.sh#L104),  install_pkgs calls [unpack_deb](https://github.com/project-stacker/c3/blob/ab7da224c682fcd6b38312cf59e16e69bde7a2bb/pkgmgr/defs.sh#L33) which downloads the packages with apt (including its dependencies) and then uses dpkg-deb to install all the downloaded packages.

Is that wrong?

Reply:
Sounds good to me, @rchincha ?

Reply:
It is super-important that we be able to do reproducible builds which means something like this:
`sudo apt-get install <package name>=<version>`

vs

`sudo apt-get install <package name>`

because if the registry is destroyed we need to reproduce images **exactly** and every version upgrade is **explicitly** tracked via PRs to c3's config files.

> in [defs.sh](https://github.com/project-stacker/c3/blob/ab7da224c682fcd6b38312cf59e16e69bde7a2bb/pkgmgr/defs.sh#L104), install_pkgs calls [unpack_deb](https://github.com/project-stacker/c3/blob/ab7da224c682fcd6b38312cf59e16e69bde7a2bb/pkgmgr/defs.sh#L33) which downloads the packages with apt (including its dependencies) and then uses dpkg-deb to install all the downloaded packages.

> Is that wrong?

Probably not, but that is already being tracked in the pkg list.

Reply:
> because if the registry is destroyed we need to reproduce images exactly and every version upgrade is explicitly tracked via PRs to c3's config files.

this approach simply will not work for ubuntu without something like http://snapshot.debian.org/archive/debian/ . And for debian, you need to use snapshot.debian.org to achieve this.

Packages "fall off" of -updates and -security in Ubuntu (I'm not sure how debian handles this).  Immediately after package version N+1 is uploaded to -updates or -security, version N will not appear in apt metadata.  So at that point apt wont "know" about the previous version, so your 'package=VERSION' install will fail.

Even if you did keep the old/stale package data around,  you will fail on Ubuntu public mirrors.  That is because at some point after version N+1 is uploaded, the version N binaries are removed from the "pool/" .. (They have been replaced , no current metadata points at them so they are garbage collected).

Reply:
> this approach simply will not work for ubuntu without something like http://snapshot.debian.org/archive/debian/ . And for debian, you need to use snapshot.debian.org to achieve this.

so be it?

Reply:
> > this approach simply will not work for ubuntu without something like http://snapshot.debian.org/archive/debian/ . And for debian, you need to use snapshot.debian.org to achieve this.
> 
> so be it?

But we need similar solutions for the other 2 distros. Not sure for Rocky, but for Centos the content of the repos also changes.

Reply:
http://dl.rockylinux.org/vault/rocky/
http://archive.ubuntu.com/ubuntu/
^ ?

Reply:
Also note that we intend to build only a few of these container images and others can derive from these.
Aligned along the lines of distroless project with better provenance.

Reply:
1. Can we merge this PR as is or after we improved the package tracking / build scripts?
2. For tracking the package versions I have the following proposal: https://github.com/project-stacker/c3/pull/88/files
- #88 only applies for the static image atm
- it can generate a list of versioned packages to be installed based on the dependencies decided by package authors
- there would be 2 separate steps, the first one is to run the script to update the package list, this can be done when developers decide it's time to make an update, the second one if the actual installation (missing from my PR atm) which would be similar to how we run the build right now
- if you think #88 would solve at least one of our problems (the repo urls can be fixed afterwards), then I will continue working on it

Reply:
> http://dl.rockylinux.org/vault/rocky/ http://archive.ubuntu.com/ubuntu/ ^ ?

Both these contain packages for older releases. Not older versions of packages which were previously available in the current release.

The rocky vault for example contains packages for 9.0 but not from 9.1. So if I want to build a 9.1 image today, I can't use vault as the download url is broken. Only 9.0 is available at vault right now. In the future when 9.2 becomes available, 9.1 would be moved to the vault, so then it would be possible to build 9.1 from the vault, an update to the git history modifying the yum configuration would be needed for rebuild.

Even for the debian snapshot. the latest date available, is not today, it is yesterday. So I can't build an image with today's version of the packages.

Reply:
https://github.com/project-stacker/c3/pull/88
^ work tracked here

Approving this.

Reply:
https://github.com/rocky-linux/peridot/issues/18
^ rocky folks still sorting out some issues